department-of-veterans-affairs/caseflow
# #2865: Add page timings

# Update
I removed the reducer tracking. GA is optimized for tracking `10ms+` timings, and the reducer often completes much faster than that. If we have a particular problematic action, we can instrument it specifically down the road.

# Original Content

Connect #2865.

I know this is not previously-planned work, so feel free to de-prioritize reviewing this. I was experimenting with Google Analytics Page Speed Timings as something that could help us, particularly with Mark refactoring PDF rendering and us wanting to know how that refactor affected performance. This PR is the result of that experimentation. We can either say it's useful and merge it, or put it on the back-burner for now.

The pieces to this are:

1. ~Update `PerfDebug.timeFunction` to accept a callback for when the timed function is complete.~
1. Add `PerfDebug.timeFunctionPromise` to time async operations.
1. ~Instrument all a throttled tracker on all Reader reducer calls.~
1. Update `ApiUtil` to track the client-side timing of perf requests.
1. Instrument all Reader HTTP requests to declare which endpoint they are hitting, for analytics.

I can split these into separate PRs if you'd prefer.

I started with Reader. If this is seen to be useful, we can roll it out more broadly. We may also have the experience of reworking this once we see what the data actually looks like in GA, in which case it would be good to have started with a smaller footprint.

## Future Timings to Track
1. PDF render times (but I don't want to interfere with @mdbenjam's ongoing refactor so this will wait)
1. "Page usefulness" times. For example, how long after the page loads does the document list fully render? How long does it take between when a user clicks on the document type link and when the document is fully rendered for the user on the next page? (Note: [this may be fairly difficult](https://stackoverflow.com/questions/46413777/knowing-when-the-ui-is-ready-in-react-redux).)
1. User behavior. We could time how long a user spends on a doc, for instance.

## Test
1. Observe GA traffic in the network tab:

![image](https://user-images.githubusercontent.com/829827/30711480-d87fccf4-9ed6-11e7-91f4-637ee3c30219.png)

We can see the information we're passing show up in the query params.

2. Look in the Google Analytics UI for the account `UA-74789258-5`, and go to **Behavior** > **User Timings** (and other timings pages).

![image](https://user-images.githubusercontent.com/829827/30830763-20b31618-a213-11e7-9de9-f55825495c36.png)

3. Ensure that page perf is not degraded by sending too many tracking HTTP requests.
